### PR TITLE
feat: implement and

### LINKED ISSUE #11: feat: implement 0x16 - AND opcode
---
fork: Frontier
group: Comparison & Bitwise Logic Operations
---

*Index 1 is top of the stack. See [PUSH](https://www.evm.codes/#60).*

## Stack input

0. `a`: first binary value.
1. `b`: second binary value.

## Stack output

0. `a & b`: the bitwise AND result.

## Examples

| * | Input | Output | * | * | Input | Output |
|--:|------:|-------:|--:|--:|------:|-------:|
| `1` | `0xF` | `0xF` | * | `1` | `0xFF` | `0` |
| `2` | `0xF` | | * | `2` | `0` | |

[Reproduce in playground](https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='z1~xF~xwyyz2~~xFw'~yPUSH1%200z%2F%2F%20Example%20y%5CnwFyAND%01wyz~_).

## Error cases

The state changes done by the current context are [reverted](https://www.evm.codes/?fork=merge#FD) in those cases:
- Not enough values on the stack.

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #11 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/tests/test_instructions/test_comparison_operations.cairo [hunk 1]
```diff
@@ -5,6 +5,67 @@ use option::OptionTrait;
 use debug::PrintTrait;
```
Comment (enitrat) [inline src/tests/test_instructions/test_comparison_operations.cairo:6]:
```suggestion
use integer::BoundedInt;
```

File: src/tests/test_instructions/test_comparison_operations.cairo [hunk 2]
```diff
@@ -5,6 +5,67 @@ use option::OptionTrait;
...
 
 
+#[test]
+#[available_gas(20000000)]
+fn test_and_zero_and_max() {
+    // Given
+    let mut ctx = setup_execution_context();
+    ctx.stack.push(0x00);
+    ctx.stack.push(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
```
Comment (enitrat) [inline src/tests/test_instructions/test_comparison_operations.cairo:14]:
```suggestion
    ctx.stack.push(BoundedInt::<u256>::max());
```

File: src/tests/test_instructions/test_comparison_operations.cairo [hunk 3]
```diff
@@ -5,6 +5,67 @@ use option::OptionTrait;
...
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    assert(ctx.stack.peek().unwrap() == 0x00, 'stack top should be 0x00');
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_and_max_and_max() {
+    // Given
+    let mut ctx = setup_execution_context();
+    ctx.stack.push(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
```
Comment (enitrat) [inline src/tests/test_instructions/test_comparison_operations.cairo:29]:
```suggestion
    ctx.stack.push(BoundedInt::<u256>::max());
```

File: src/tests/test_instructions/test_comparison_operations.cairo [hunk 4]
```diff
@@ -5,6 +5,67 @@ use option::OptionTrait;
...
+    ctx.stack.push(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
```
Comment (enitrat) [inline src/tests/test_instructions/test_comparison_operations.cairo:30]:
```suggestion
    ctx.stack.push(BoundedInt::<u256>::max());
```

File: src/tests/test_instructions/test_comparison_operations.cairo [hunk 5]
```diff
@@ -5,6 +5,67 @@ use option::OptionTrait;
...
+    // When
+    ctx.exec_and();
+
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    assert(
+        ctx
+            .stack
+            .peek()
+            .unwrap() == 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
```
Comment (enitrat) [inline src/tests/test_instructions/test_comparison_operations.cairo:41]:
```suggestion
            .unwrap() == BoundedInt::<u256>::max(),
```

File: src/tests/test_instructions/test_comparison_operations.cairo (hunk 2 above)
Comment (ftupas) [inline src/tests/test_instructions/test_comparison_operations.cairo:14]: I was thinking for tests, maybe we can use [`min()`](https://github.com/starkware-libs/cairo/blob/main/corelib/src/integer.cairo#L1426) and [`max()`](https://github.com/starkware-libs/cairo/blob/main/corelib/src/integer.cairo#L1430) so we won't accidentally miss any hex 🤔 What do you think?

**Review by ftupas — COMMENTED** [review]
Nice work! Just a question on using min/max instead of typing the literals for tests

**Review by enitrat — APPROVED** [review]

**Pull request merged**